### PR TITLE
♻️ [Button] Incompatible prop groups

### PR DESCRIPTION
**Pull request by beefchimi**
> **Important!** This PR is an experiment

Often times, we will define a `Props` interface that contains some props that will never be used together.

### Problem

In the example of our `Button` component, we want the consumer to provide either a `url` or an `onClick`.
- We won't ever utilize both of these props at once.
- In order to allow for one-or-the-other, we set them both to `optional (?)` and trust the consumer to understand the API.

The usual tactic is to single out one of the props as having higher priority - in case both are provided. In the case of `Button`:
1. We first check for the `url` prop
2. If present, render _something_ based on that condition
3. Otherwise, we pursue a different condition and hope that `onClick` was provided instead

This gets even trickier when there are "dependant props". Imagine a series of props that are only relevant in the `url` condition - such as `external` or `download`.

### Desire

It would be nice if we could more easily separate out these compatible prop groupings, and provide feedback to the developer when they are providing a prop that is not compatible / missing props that are relevant.

### Solution

We can actually define [compatible/incompatible type groupings](https://user-images.githubusercontent.com/643944/116140715-b87dd100-a6a5-11eb-8c20-5f1179ad66d2.png). This PR is a quick and dirty example of having an interface specifically for `url`, and a separate interface for `onClick`. _Credit goes to Clauderic for showing me this technique._

**Some examples of usage:**

```tsx
// Fail: Does not have url or onClick
<Button />

// Pass: has url
<Button url="https://google.ca/" />

// Pass: has onClick
<Button onClick={onDismiss} />

// Fail: external is not compatible with onClick
<Button onClick={onDismiss} external />

// Pass: external is compatible with url
<Button url="https://google.ca/" external />
```

https://user-images.githubusercontent.com/643944/116140722-ba479480-a6a5-11eb-8d47-4c1876497819.mp4

I currently have an [experimental Draft PR open elsewhere](https://github.com/Shopify/online-store-ui/pull/2754) that explores this same technique. All the various `type/interface` definitions can get a little cumbersome. Part of my exploration here is to decide if that is a worthwhile trade-off for the better type handling on the consumer side.

### Alternative solution

Another way of solving this _could look something like this_ (code untested):

<details>
<summary><strong>Example code:</strong></summary>

```tsx

enum ButtonType {
  Action = 'action',
  Link = 'link',
}

interface ButtonCommonProps {
  content: string;
  id?: string;
}

interface ButtonActionProps extends ButtonCommonProps {
  type: ButtonType.Action;
  onClick(): void;
  ariaControls?: string;
}

interface ButtonLinkProps extends ButtonCommonProps {
  type: ButtonType.Link;
  url: string;
  external?: boolean;
}

type ButtonProps = ButtonActionProps | ButtonLinkProps;

export function Button({
  type,
  id,
  content,
  onClick,
  ariaControls,
  url,
  external,
}: ButtonProps) {
  return type === ButtonType.Action (
    <button type="button" id={id} aria-controls={ariaControls} onClick={onClick}>
      {content}
    </button>
  ) : (
    <a id={id} href={url} external={external}>{content}</a>
  );
}

```

</details>

### Ask for reviewers

What are your thoughts? Is this a pattern we should move forward with?

### CODE REVIEW

**Review by clauderic — COMMENTED** [review]
Overall I'm a fan of this approach. It's worth noting that this won't work with some older versions of TypeScript, but that's probably okay. 

I wish the TypeScript error messaging were a bit less cryptic though, might make it a bit challenging for consumers to understand what's wrong. 

The proposed changes here seem like a breaking change unless I'm misunderstanding, not sure if that was intentional or accidental.

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -263,25 +269,28 @@ export function Button({
...
+        ariaDescribedBy,
+        pressed,
+        onKeyDown,
+        onKeyUp,
+        onKeyPress,
+      };
 
   const buttonMarkup = (
-    <UnstyledButton {...commonProps} {...linkProps} {...actionProps}>
+    <UnstyledButton {...commonProps} {...conditionalProps}>
```
[the remark this answers is not in the document]
Comment (beefchimi) [reply src/components/Button/Button.tsx:293]: Yes - there will be a breaking change if we go ahead with this code.

As far as the link vs action props though... while they were being passed through to `UnstyledButton`, that component has  a `if (url)` condition that effectively does the "prop sorting". So, in other words, you don't actually end up with a result that has both a `href + onClick`.

Regardless, the typing will fail if consumers are attempting to pass both at once, so it is indeed a breaking change.

[the remark this answers is not in the document]
Comment (clauderic) [reply src/components/Button/Button.tsx:293]: Nevermind, I read the code wrong, the amount of prop spreading in [UnstyledButton](https://github.com/Shopify/polaris-react/blob/main/src/components/UnstyledButton/UnstyledButton.tsx) is a bit hard to follow and wrap your head around 😅

[the remark this answers is not in the document]
Comment (beefchimi) [reply src/components/Button/Button.tsx:293]: Clauderic and I chatted a bit over Slack and it actually looks like `onClick` _could be_ getting passed through from `Button > UnstyledButton > UnstyledLink`, so I should double-check and major sure I know exactly what props make it through.

[the remark this answers is not in the document]
Comment (BPScott) [reply src/components/Button/Button.tsx:293]: Shouldn't we already throw an error if you pass both url and onclick into Button in the first place? And thus at this point we know that there's never going to be  both at the same time, so we don't need to check it what we pass through to unstyledbutton?

**Comment by BPScott** [comment]
Polaris uses this idea in a whole two places already, so there's prior art for it, it's just never been prioritised, as it is strictly a breaking change - code like `<Button url="" onClick={()=>{}}>` previously passed but will now fail type checking.

- [Textfield](https://github.com/Shopify/polaris-react/blob/0ab82330d35444c6d7f19c303b200c22c9340cbd/src/components/TextField/TextField.tsx#L135-L140)
- [Tag](https://github.com/Shopify/polaris-react/blob/main/src/components/Tag/Tag.tsx#L22-L26)


I don't think you need the 

```ts
  // Incompatible props:
  url?: never;
  external?: never;
  download?: never;
```

stuff because the type should already [discriminated](https://basarat.gitbook.io/typescript/type-system/discriminated-unions) by the presence of either `url` (and thus you have an object of type ButtonUrlProps) or `onClick`  (and thus you have a type of ButtonActionProps).

Question: Do we need the BaseButton type? Having it be a partial base for two components feels confusing. Would it improve readability to have an UnstyledButtonProps type and have ButtonProps extend from UnstyledButton? That `[key: string]: any;` looks like it'd mess that up, why's that needed?


---

Possible refutation:

Is `<Button url="" onClick={()=>{}}>` actual valid code? In my old non-reacty days I remember plenty of progressive enhancement cases where `<a href="BlAH" onClick="doThing">` was desired behaviour - clicking on the link when JS was loaded would give you a modal or something and middle-click / open-in-new-tab / clicking before JS was loaded would take you to a new page.
